kikit/fab/jlcpcb.py

```diff
@@ -3,30 +3,64 @@
 import csv
 import os
 import shutil
+from math import sin, cos, radians
 from pathlib import Path
 from kikit.eeshema import extractComponents, getField
 from kikit.defs import MODULE_ATTR_T
 from kikit.fab.common import hasNonSMDPins
 from kikit.common import *
 from kikit.export import gerberImpl
 
+class FormatError(Exception):
+    pass
+
 def layerToSide(layer):
     if layer == pcbnew.F_Cu:
         return "T"
     if layer == pcbnew.B_Cu:
         return "B"
     raise RuntimeError(f"Got component with invalid layer {layer}")
 
-def moduleX(module, placeOffset):
+def modulePosition(module, placeOffset, compensation):
     pos = module.GetPosition() - placeOffset
+    angle = -radians(module.GetOrientation() / 10.0)
```

Comment:
Maybe I'm not thinking through this correctly, but if you're making a rotation correction and an offset correction, is it correct to apply the rotation before offsetting the center of rotation like this?  I'm having trouble visualizing precisely how I might use the x,y correction feature, but I *think* it would be to bring an incorrect part's origin into the parts actual center of rotation.  I would probably want to apply any rotation offset after that?

Reply:
The use case for XY correction comes from solder jumpers. In some of my designs, I use solder jumpers - a 3 pad component where you can place zero-ohm resistor to two places. The solder jumper has the origin in its center. By using 0402 resistors and by applying the correction I can select which configuration will be populated by the assembly house.

Reply:
Hmm, I see.  I don't know what facilities KiCAD makes available, but personally I would want something like this visible from the schematic.  In the past, I've used the solder jumpers which is not a BOM item and can be "programmed" through selecting the right footprint.  Personally, specifying an offset in the `JCLPCB_CORRECTION` (manufacturer specific) to create a connectivity change seems to mix multiple concerns in a rather opaque way that I would tend to avoid, but maybe this is the only way based on your workflow?  In any case, I think it's important to note in the documentation the order the correction is applied (rotate, then translate).

Reply:
The problem with having multiple footprints and changing them is that they have different origins and your footprint will move when you change it, therefore, you cannot automate it. I would like to easily produce different configurations of the boards. When I think about it, it would be nice to have an option to specify list of fields with the corrections where the first found is used. Then you can have fields like `JLCPCB_CORRECTION` for all components and for some `JLCPCB_CORRECTION_CFG1` and `JLCPCB_CORRECTION_CFG2`. This way it is transparent what the offset means.

Also one note for not going the footprint way - the connectivity is the matter of assembly more than board layout. With a different manufacturer, you might want to go a different way.

You are right about the documentation. I'll make sure I put some examples there.

Reply:
Not trying to convince you of anything, but just to clarify what I meant: changing the footprint to "reprogram" the jumper was in the case of the solder jumpers.  These can have the same origin and rotation, but with a footprint with option A connected or option B connected (in copper).  These are just defaults, and can be reprogrammed by cutting the trace in the jumper and solder bridging the two other contacts.  I see the difficulty in doing this if you want to populate 0 Ohm Resistors, I was just trying to propose an alternate approach.

I also don't think I agree with your approach to connectivity.  To me, all the connectivity should be captured in the schematic, since that is ultimately what the schematic is for.  I think when you change manufacturers in KiKit, you should get the same design just targeted to a different manufacturer.  IMO, which jumper options are selected should be captured in the design somewhere, not in the translation from design to manufacturing package.

Thinking about it a bit more, I think there are other problems with the approach too.  If you have a design with 3 jumpers with the same footprint, I don't see how you can independently change what each is programmed to (JP1=A, JP2=B, and so on)?

I hope this comment didn't read as overly critical, I don't mean to argue against your workflow, especially since I don't have a strong alternative to offer.  However, I am still unconvinced that using the per-manufacturer placement translation in this way is a good idea and wanted to give you my reasons so you can consider them.

Reply:
I don't take it critically and I am open to other opinions and views.

IMHO the connectivity is captured in the schematics - it is a jumper after all. I just program it to a default value (possibly based on a text note in schematics). The corrections are per instance of a component, not per footprint, so you can easily do JP1=A, JP2=B.

Reply:
I think I see one of the disconnects.  To me, I would save the manufacturer corrections back to my library.  That way in the future, if I use that part it already has an `LCSC part number`, `JLCPCB_CORRECTION` etc. assigned.  This is pretty incompatible with tweaking `JLCPCB_CORRECTION` to reprogram jumpers.  It sounds like your approach is to tweak all the corrections anyway, so what is baked in as the default is less of an issue for you.  Translating text notes to offsets and repeating this for every manufacturer you want to support, for every jumper in every design still seem tedious and error prone to me, but maybe it's the only way.

Anyway, it's easy enough for me to just not use the translation feature so I wouldn't take my opinion as a reason to block merging this :smile: